### PR TITLE
Chunkwise iteration over arrays.

### LINKED ISSUE #398: Default efficient row iteration
It seems like iterating over a chunked array is inefficient at the moment, presumably because we're repeatedly decompressing the chunks. For example, if I do
```python
    for pos, row in enumerate(data_root.variants):
        print(row)
        if pos == 1000:
            break
```
it takes several minutes (data_root.variants is a large 2D chunked matrix) but if I do
```python
    for pos, row in enumerate(chunk_iterator(data_root.variants)):
        print(row)
        if pos == 1000:
            break
```
it takes less than a second, where

```python
def chunk_iterator(array):
    """ 
    Utility to iterate over the rows in the specified array efficiently
    by accessing one chunk at a time.
    """
    chunk_size = array.chunks[0]
    for j in range(array.shape[0]):
        if j % chunk_size == 0:
            chunk = array[j: j + chunk_size][:]
        yield chunk[j % chunk_size]
```

To me, it's quite a surprising gotcha that zarr isn't doing this chunkwise decompression, and I think it would be good to do it by default. There is a small extra memory overhead, but I think that's probably OK, given the performance benefits.

Any thoughts?

### PR DESCRIPTION
**Pull request by jeromekelleher**
Closes #398.

Thanks to @alimanfoo for giving me the formula for unit tests! The only tricky thing is the zero-d case, which I've mirrored numpy's behaviour for.

TODO:
* [x] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [x] Changes documented in docs/release.rst
* [ ] Docs build locally (e.g., run ``tox -e docs``)
* [x] AppVeyor and Travis CI passes
* [x] Test coverage is 100% (Coveralls passes)

### CODE REVIEW

File: zarr/core.py [hunk 1]
```diff
@@ -424,6 +424,18 @@ def __array__(self, *args):
...
+    def __iter__(self):
+        if len(self.shape) == 0:
+            # Same error as numpy
+            raise TypeError("iteration over a 0-d array")
+        # Avoid repeatedly decompressing chunks by iterating over the chunks
+        # in the first dimension.
+        chunk_size = self.chunks[0]
+        for j in range(self.shape[0]):
+            if j % chunk_size == 0:
+                chunk = self[j: j + chunk_size][:]
```
[the remark this answers is not in the document]
Comment (jeromekelleher) [reply zarr/core.py]: I thought you would need to take an extra copy to be sure the data was kept in a local numpy array. Changed now.

**Comment by jeromekelleher** [comment]
Thanks @alimanfoo, I've implemented those. **Love** the changelog format btw, I think I'll be copying that!

File: zarr/core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alimanfoo) [reply zarr/core.py]: Thanks, yeah just the single slice `[j: j + chunk_size]` is enough to bring data into a local numpy array. Adding the extra `[:]` is harmless but not necessary.

File: docs/release.rst [hunk 2]
```diff
@@ -23,15 +23,18 @@ Enhancements
   SQLite database to be used as the backing store for an array or group.
   By :user:`John Kirkham <jakirkham>`, :issue:`368`, :issue:`365`.
 
+* Efficient iteration over arrays by decompressing chunkwise.
+  By :user:`Jerome Kelleher <jeromekelleher>`, :issue:`398`.
```
Comment (alimanfoo) [inline docs/release.rst]:
```suggestion
  By :user:`Jerome Kelleher <jeromekelleher>`, :issue:`398`, :issue:`399`.
```

Comment (alimanfoo) [reply docs/release.rst]: Tiny thing, I'll resolve later.

Comment (jeromekelleher) [reply docs/release.rst]: Cool --- wasn't sure if this was referring to the issue or the PR  as well.

**Pull request merged**